p2p/protocols/protocol.go

```diff
@@ -277,15 +278,31 @@ func (p *Peer) Send(ctx context.Context, msg interface{}) error {
 		}
 		size = len(r)
 	}
-	// if the accounting hook is set, call it
+
+	// if the accounting hook is set, do accounting logic
 	if p.spec.Hook != nil {
-		err = p.spec.Hook.Send(p, uint32(size), msg)
+
+		// let's lock, we want to avoid that after validating, a separate call might interfere
+		p.lock.Lock()
+		defer p.lock.Unlock()
+		// validate that this operation would succeed...
+		costToLocalNode, err := p.spec.Hook.Validate(p, uint32(size), wmsg, Sender)
```

Comment:
If this function is called concurrently, multiple `amounts` could be "validated" (none of those amounts are applied), but when applied, the majority could still fail.

Reply:
No, my bad. This is not possible as we are under a lock!